src/types.rs

```diff
@@ -475,3 +506,150 @@ impl<'buf> RecvMsgOut<'buf> {
+///
+/// The default behavior of the [CancelBuilder] is cancel all in-flight requests,
+/// however this can be refined to cancel only requests which reference a file descriptor
+/// or have a specific `user_data` value.
+///
+/// Additionally, [CancelBuilder::all] can be set to match multiple requests.
+///
+/// ```
+/// use io_uring::types::{CancelBuilder, Fd, Fixed};
+///
+/// // Match all in-flight requests.
+/// CancelBuilder::new();
+///
+/// // Match a single request with user_data = 42.
+/// CancelBuilder::new().user_data(42);
+///
+/// // Match a single request with fd = 42.
+/// CancelBuilder::new().fd(Fd(42));
+///
+/// // Match a single request with fixed fd = 42.
+/// CancelBuilder::new().fd(Fixed(42));
+///
+/// // Match all in-flight requests with user_data = 42.
+/// CancelBuilder::new().user_data(42).all();
+/// ```
+#[derive(Debug)]
+pub struct CancelBuilder {
+    pub(crate) flags: AsyncCancelFlags,
+    pub(crate) user_data: Option<u64>,
+    pub(crate) fd: Option<sealed::Target>,
+}
+
+impl CancelBuilder {
+    /// Create a new [CancelBuilder].
+    ///
+    /// The default behavior of the [CancelBuilder] is cancel all in-flight requests.
+    /// This behavior can be further refined by calling the builder methods on the [CancelBuilder].
+    pub fn new() -> Self {
+        Self {
+            flags: AsyncCancelFlags::ANY,
+            user_data: None,
+            fd: None,
+        }
+    }
+
+    /// Refine the match criteria to match only requests which contain the
+    /// provided `user_data`.
+    ///
+    /// This is mutually exclusive with [CancelBuilder::fd](#method.fd).
+    pub fn user_data(mut self, user_data: u64) -> Self {
+        // Unset the ANY flag because we want to refine the match criteria to a subset of requests.
+        self.flags.set(AsyncCancelFlags::ANY, false);
+        // Unset the FD flags, as they are mutually exclusive with user_data.
+        self.flags.set(AsyncCancelFlags::FD, false);
+        self.flags.set(AsyncCancelFlags::FD_FIXED, false);
+        self.fd = None;
+
+        self.user_data = Some(user_data);
+        self
+    }
+
+    /// Refine the match criteria to match only requests which reference the
+    /// provided `fd`.
+    ///
+    /// This is mutually exclusive with [CancelBuilder::user_data](#method.user_data).
+    ///
+    /// Note: Support for fixed file descriptors is only available on Linux 6.0+.
+    pub fn fd(mut self, fd: impl sealed::UseFixed) -> Self {
+        // Unset the ANY flag because we want to refine the match criteria to a subset of requests.
+        self.flags.set(AsyncCancelFlags::ANY, false);
+        let target = fd.into();
+        self.user_data = None;
+        match target {
+            sealed::Target::Fd(_) => {
+                self.flags.set(AsyncCancelFlags::FD, true);
+            }
+            sealed::Target::Fixed(_) => {
+                self.flags.set(AsyncCancelFlags::FD_FIXED, true);
```

Comment:
For this, I would have thought both need to be set, but if you've tested and it works this way, then that's fine.

Reply:
Thanks for pointing this out! I checked again, I think you are right that we need both CANCEL_FD and CANCEL_FD_FIXED.

The checks for CANCEL_FD_FIXED seem to be gated on CANCEL_FD.
https://github.com/torvalds/linux/blob/3a93e40326c8f470e71d20b4c42d36767450f38f/io_uring/cancel.c#L177-L179

I updated the PR to include a test for fixed fd cancellation, and it fails unless we include the CANCEL_FD_FIXED flag.